### DISCUSSION TITLE
brew services (coredns) service path root ownership breaks brew

### DISCUSSION THREAD

**Discussion by af3556** [Homebrew/discussions#3854] [Everyday usage] [answered]
Starting coredns as a service chowns some paths to root:

    % sudo brew services start coredns
    Warning: Taking root:admin ownership of some coredns paths:
      /opt/homebrew/Cellar/coredns/1.9.3/bin
      /opt/homebrew/Cellar/coredns/1.9.3/bin/coredns
      /opt/homebrew/opt/coredns
      /opt/homebrew/var/homebrew/linked/coredns
    This will require manual removal of these paths using `sudo rm` on
    brew upgrade/reinstall/uninstall.
    ==> Successfully started `coredns` (label: homebrew.mxcl.coredns)

Subsequent brew operations then fail:

    % brew ...
    ...
    Error: An unexpected error occurred during the `brew link` step
    The formula built, but is not symlinked into /opt/homebrew
    Permission denied @ rb_readlink - /opt/homebrew/opt/coredns
    Error: Permission denied @ rb_readlink - /opt/homebrew/opt/coredns

Based on the warning emitted by the services start, seems like this is working as expected; it's not terribly friendly though.

At the least, it seems preferable to update the warning to a chmod of the offending paths rather than remove them.

Is there really no other way than remembering to do the above every time I use brew after starting coredns as a background service?

**Comment by af3556** [comment]
Everything seems to work ok if I run coredns as a user, via launchd, rather than as a system process via `brew services`.

    % sudo brew services stop coredns
    % sudo launchctl bootstrap user/$(id -u) $(brew list coredns | grep plist)
    % launchctl print user/$(id -u)/homebrew.mxcl.coredns

**Reply by SMillerDev** [reply]
Then why not just run brew service without sudo?

**Reply by af3556** [reply]
doh

Indeed, that does the trick. Face palm as to why I didn't consider this earlier; for some reason had the mistaken idea that brew services would only operate on system-wide services, i.e. /Library/LaunchDaemons/.

**Comment by af3556** [comment, marked as answer]
To avoid issues with launchd's policy where globally installed daemons must be owned by the root user, just run the service as the logged in user:

    % brew services start coredns
    Warning: coredns must be run as root to start at system startup!
    ==> Successfully started `coredns` (label: homebrew.mxcl.coredns)

i.e. don't use `sudo`

The service will be started when the user logs in.

If you need it running when the user is not logged in, you could DIY the launchd setup in the `user/` launchd domain, as above.

**Reply by Salhazry** [reply]
git remote add origin https://github.com/Salhazry/My-saif.git
 git Branch -M main 
git push -u origin main

**Reply by Salhazry** [reply]
> git remote add origin https://github.com/Salhazry/My-saif.git
>  git Branch -M main 
> git push -u origin main
